### DISCUSSION TITLE
Macroable Panel

### DISCUSSION THREAD

**Discussion by ronssij** [filamentphp/filament#13883] [Ideas]
@danharrin 
I suggest that `Filament/Panel` should be a `Macroable` class.

If project(s) jas modularized features you will register `Resources, Pages etc.` like so:
![image](https://github.com/user-attachments/assets/5e4a8769-a6e2-4b26-bfac-d08cc105e3b9)

Currently this is acheivable dynamically like so:
![image](https://github.com/user-attachments/assets/dba53439-1197-4651-86be-7c610e11bc88)

If we use macroable
```php
Panel::macro('discoverModules', function () {
    concord()->getModules()->each(function (BaseBoxServiceProvider $module) {
        $basePath = $module->getBasePath();
        $namespaceRoot = $module->getNamespaceRoot();

        $this->discoverResources(
            in: "$basePath/Filament/Resources",
            for: "$namespaceRoot\\Filament\\Resources",
        )->discoverPages(
            in: "$basePath/Filament/Pages",
            for: "$namespaceRoot\\Filament\\Pages",
        )->discoverWidgets(
            in: "$basePath/Filament/Widgets",
            for: "$namespaceRoot\\Filament\\Widgets",
        )->discoverLivewireComponents(
            in: "$basePath/Filament/Livewire",
            for: "$namespaceRoot\\Filament\\Livewire",
        );
    });
})
```  
 Then on panel provider.
```php        
return $panel
      ->login()
      ->discoverModules()
```

**Comment by danharrin** [comment]
Judging by the code, it already is Macroable. What error happens when you attempt to macro? What version are you using?

**Reply by ronssij** [reply]
Is it because the `Component` class us Macroable?

```
![image](https://github.com/user-attachments/assets/a10b57dd-6d2d-43a6-b4a7-a262a7d34128)

```

But it seems it does not work.
![image](https://github.com/user-attachments/assets/506cf09c-f317-4462-961d-8feed9c2755d)

**Reply by ronssij** [reply]
@danharrin It does work. I shouldve put it on the `register()` method on AppServiceProvider
